## **Pandas Essencial**
**Prof. Dr. Samuel Martins (@hisamuka @xavecoding)** <br/>
xavecoding: https://youtube.com/c/xavecoding <br/>

Neste tutorial, vamos aprender o essencial da biblioteca _Pandas_ para manipulação de dados.<br/><br/>

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.

### Pacotes usados neste Notebook

In [116]:
# pacotes usados neste notebook
import pandas as pd

In [117]:
# Conecta com o google drive para extrair a base de dados
from google.colab import drive
drive.mount('/content/drive')

# Caminho padrao para a base de dados
path = '/content/drive/MyDrive/trilia-dados/'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


<h1>Manipulação de Dados com Pandas</h1><hr/>

<h2>1. Manipulação Básica de Datasets</h2>
<hr/>

**Dataset**: Gas Prices in Brazil: https://www.kaggle.com/matheusfreitag/gas-prices-in-brazil <br/>

Este dataset contém os **registros dos preços médios semanais dos combustíveis do Brasil entre os anos de 2004 e 2019**. <br/>
Cada *amostra (registro/linha)* consiste em um registro de preço aferido para um dado tipo de combustível em uma dada localidade do Brasil. <br/>
Alguns dos principais *atributos* (colunas) do dataset são: 'ESTADO', 'PRODUTO', 'NÚMERO DE POSTOS PESQUISADOS', 'PREÇO MÉDIO REVENDA'.


\* O arquivo disponibilizado no Kaggle está no formato *tsv*. Embora o _pandas_ consiga abrí-lo normalmente, convertemos tal arquivo para o formato *CSV*, que é um dos formatos mais utilizados, e mudamos seu separado para ';' apenas para mostrar algumas opções da função de carregamento.

### 1.1. Importando o Dataset
Para carregar um dataset no formato csv, basta utilizar a função `read_csv` do pandas. Por padrão, ela considera _','_ como separador.

In [118]:
data = pd.read_csv(path + "GasPricesinBrazil_2004-2019.csv")

In [119]:
from typing_extensions import dataclass_transform
data

,Unnamed: 0;DATA INICIAL;DATA FINAL;REGIÃO;ESTADO;PRODUTO;NÚMERO DE POSTOS PESQUISADOS;UNIDADE DE MEDIDA;PREÇO MÉDIO REVENDA;DESVIO PADRÃO REVENDA;PREÇO MÍNIMO REVENDA;PREÇO MÁXIMO REVENDA;MARGEM MÉDIA REVENDA;COEF DE VARIAÇÃO REVENDA;PREÇO MÉDIO DISTRIBUIÇÃO;DESVIO PADRÃO DISTRIBUIÇÃO;PREÇO MÍNIMO DISTRIBUIÇÃO;PREÇO MÁXIMO DISTRIBUIÇÃO;COEF DE VARIAÇÃO DISTRIBUIÇÃO;MÊS;ANO
0,0;2004-05-09;2004-05-15;CENTRO OESTE;DISTRITO ...
1,1;2004-05-09;2004-05-15;CENTRO OESTE;GOIAS;ETA...
2,2;2004-05-09;2004-05-15;CENTRO OESTE;MATO GROS...
3,3;2004-05-09;2004-05-15;CENTRO OESTE;MATO GROS...
4,4;2004-05-09;2004-05-15;NORDESTE;ALAGOAS;ETANO...
...,...
106818,106818;2019-06-23;2019-06-29;NORDESTE;RIO GRAN...
106819,106819;2019-06-23;2019-06-29;SUL;RIO GRANDE DO...
106820,106820;2019-06-23;2019-06-29;SUL;SANTA CATARIN...
106821,106821;2019-06-23;2019-06-29;SUDESTE;SAO PAULO...


O dataset não foi carregado corretamente pois o separador utilizado seu arquivo CSV era ';' e não a ','. <br/>
Vamos então carregá-lo corretamente:

In [120]:
data = pd.read_csv(path + "GasPricesinBrazil_2004-2019.csv",
                   sep = ";")

In [121]:
data

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,3.690,0.818,0.018,2.756,0,2.756,2.756,0,6,2019
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,3.789,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,3.499,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,3.490,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019


### 1.2. Exibindo as primeiras linhas do Dataset
A função `.head()` exibe as 5 primeiras linhas do dataset/tabela/Data Frame.

In [122]:
data.head(10)

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
5,5,2004-05-09,2004-05-15,NORDESTE,BAHIA,ETANOL HIDRATADO,408,R$/l,1.383,0.132,...,2.050,0.426,0.095,0.957,0.128,0.5686,1.35,0.134,5,2004
6,6,2004-05-09,2004-05-15,NORDESTE,CEARA,ETANOL HIDRATADO,278,R$/l,1.453,0.218,...,1.950,0.353,0.150,1.1,0.068,0.7332,1.294,0.062,5,2004
7,7,2004-05-09,2004-05-15,NORDESTE,MARANHAO,ETANOL HIDRATADO,105,R$/l,1.631,0.158,...,2.000,0.515,0.097,1.116,0.091,0.9923,1.36183,0.082,5,2004
8,8,2004-05-09,2004-05-15,NORDESTE,PARAIBA,ETANOL HIDRATADO,125,R$/l,1.284,0.130,...,1.699,0.353,0.101,0.931,0.072,0.7577,1.2972,0.077,5,2004
9,9,2004-05-09,2004-05-15,NORDESTE,PERNAMBUCO,ETANOL HIDRATADO,423,R$/l,1.224,0.141,...,1.700,0.277,0.115,0.947,0.091,0.5643,1.3541,0.096,5,2004


### 1.3 Informações do Dataset e Elementos Chave

### 1.3.1 Informações gerais sobre o Dataset

In [123]:
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106823 entries, 0 to 106822
Data columns (total 21 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Unnamed: 0                     106823 non-null  int64  
 1   DATA INICIAL                   106823 non-null  object 
 2   DATA FINAL                     106823 non-null  object 
 3   REGIÃO                         106823 non-null  object 
 4   ESTADO                         106823 non-null  object 
 5   PRODUTO                        106823 non-null  object 
 6   NÚMERO DE POSTOS PESQUISADOS   106823 non-null  int64  
 7   UNIDADE DE MEDIDA              106823 non-null  object 
 8   PREÇO MÉDIO REVENDA            106823 non-null  float64
 9   DESVIO PADRÃO REVENDA          106823 non-null  float64
 10  PREÇO MÍNIMO REVENDA           106823 non-null  float64
 11  PREÇO MÁXIMO REVENDA           106823 non-null  float64
 12  MARGEM MÉDIA REVENDA          

A primeira coluna da tabela, chamada _'Unnamed: 0'_, parece não significar nada. Na verdade, ela parece ser os **índices** da tabela, que foram salvos como uma _coluna_.<br/>
Veremos jajá como **remover tal coluna**.

Outro ponto é que, aparentemente, nenhum atributo/coluna possui valores nulos (_null_), uma vez que o número de registros do dataframe e os números de valores _non-null_ é de **106823**. <br/>
Mas, veremos que não é bem assim para esse caso.

### 1.3.2 Data Frame
Todo dataset carregado (dados estruturados) é um `Data Frame`: 'Tabela' bi-dimensional, de tamanho mutável, com dados potencialmente heterogêneos. <br/>

In [124]:
data.shape

(106823, 21)

Podemos acessar as **dimensões do Data Frame** (número de linhas x número de colunas) utilizando o atributo `.shape` do Data Frame.

In [125]:
print(f'o dataframe possui {data.shape[0]} linhas/atributos/registros e {data.shape[1]} colunas/atributos/variaveis  ')

o dataframe possui 106823 linhas/atributos/registros e 21 colunas/atributos/variaveis  


#### **Criando um DataFrame**

Podemos criar um DataFrame a partir de um _dicionário_, onde cada **chave** possui uma **lista de elementos de igual tamanho**.<br/>
As **chaves** representam as **colunas** e **cada um dos valores de sua lista** representa o **valor da linha** correspondente para aquela coluna.

In [126]:
alunos_df = pd.DataFrame({
    'nome': ['Luke Skywalker', 'Yoda', 'Palpatine'],
    'idade': [16, 1000, 70],
    'peso': [70, 15, 60],
    'eh jedi': [True, True, False]  # o nome das colunas podem ter espaços
})

In [127]:
alunos_df


,nome,idade,peso,eh jedi
0,Luke Skywalker,16,70,True
1,Yoda,1000,15,True
2,Palpatine,70,60,False


#### **VEJA MAIS**
Criando um Data Frame a partir de um dicionário: https://www.geeksforgeeks.org/how-to-create-dataframe-from-dictionary-in-python-pandas/

#### **Renomeando as colunas de um DataFrame**
**===>** O método `DataFrame.columns` retorna uma ";lista" com os **nomes de todas as colunas** do data frame.

In [128]:
alunos_df.columns

Index(['nome', 'idade', 'peso', 'eh jedi'], dtype='object')

In [129]:
type(alunos_df.columns)

pandas.core.indexes.base.Index

In [130]:
list(alunos_df.columns)

['nome', 'idade', 'peso', 'eh jedi']

<br/>

**===>** Para **renomear colunas** do data frame, utilize o método `DataFrame.rename`, que retorna uma _cópia_ do data frame com as as colunas renomeadas:

In [131]:
alunos_df

,nome,idade,peso,eh jedi
0,Luke Skywalker,16,70,True
1,Yoda,1000,15,True
2,Palpatine,70,60,False


In [132]:
alunos_df.rename(columns={
    "nome":" Nome Completo",
    "idade": "Idade Atual"
})

,Nome Completo,Idade Atual,peso,eh jedi
0,Luke Skywalker,16,70,True
1,Yoda,1000,15,True
2,Palpatine,70,60,False


Para renomear o _próprio_ data frame em questão, utilize o parâmetro `inplace=True`:

<br/>

**===>** Uma outra forma de **renomear todas as listas** de um data frame é passar uma _lista_ com os novos nomes das colunas para o atributo `DataFrame.columns`:

### 1.3.3 Series

Array uni-dimensional com os dados e rótulos de um eixo.

In [133]:
data.head()

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [134]:
data["ESTADO"]

,ESTADO
0,DISTRITO FEDERAL
1,GOIAS
2,MATO GROSSO
3,MATO GROSSO DO SUL
4,ALAGOAS
...,...
106818,RIO GRANDE DO NORTE
106819,RIO GRANDE DO SUL
106820,SANTA CATARINA
106821,SAO PAULO


In [135]:
data.ESTADO

,ESTADO
0,DISTRITO FEDERAL
1,GOIAS
2,MATO GROSSO
3,MATO GROSSO DO SUL
4,ALAGOAS
...,...
106818,RIO GRANDE DO NORTE
106819,RIO GRANDE DO SUL
106820,SANTA CATARINA
106821,SAO PAULO


In [136]:
type(data["ESTADO"])

pandas.core.series.Series

In [137]:
data.iloc[1]

,1
Unnamed: 0,1
DATA INICIAL,2004-05-09
DATA FINAL,2004-05-15
REGIÃO,CENTRO OESTE
ESTADO,GOIAS
PRODUTO,ETANOL HIDRATADO
NÚMERO DE POSTOS PESQUISADOS,387
UNIDADE DE MEDIDA,R$/l
PREÇO MÉDIO REVENDA,1.162
DESVIO PADRÃO REVENDA,0.114


In [138]:
type(data.iloc[1])

pandas.core.series.Series

#### **Criando uma Series**

Podemos criar um DataFrame a partir de uma lista de elementos.

In [139]:
pd.Series([5.5,9.9,1.0])

,0
0,5.5
1,9.9
2,1.0


Podemos alterar o **nome dos índices** (veremos melhor jajá) e o **nome da Series** (o que ela representa):

In [140]:
pd.Series([5.5,9.9,1.0],index=['prova1','prova2','prova3'],name= 'Notas dos darth vader')

,Notas dos darth vader
prova1,5.5
prova2,9.9
prova3,1.0


#### **VEJA MAIS**
https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html

### 1.3.4 Atribuindo Dados

#### 1.3.4.1 Atribuindo constantes

In [141]:
produto_view= data.head()
produto_view

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [142]:
data.PRODUTO

,PRODUTO
0,ETANOL HIDRATADO
1,ETANOL HIDRATADO
2,ETANOL HIDRATADO
3,ETANOL HIDRATADO
4,ETANOL HIDRATADO
...,...
106818,GNV
106819,GNV
106820,GNV
106821,GNV


In [143]:
produto_copy=data.PRODUTO.copy()
produto_copy

,PRODUTO
0,ETANOL HIDRATADO
1,ETANOL HIDRATADO
2,ETANOL HIDRATADO
3,ETANOL HIDRATADO
4,ETANOL HIDRATADO
...,...
106818,GNV
106819,GNV
106820,GNV
106821,GNV


In [144]:
data['PRODUTO'] = 'COMBUSTIVEL' #todos itens da coluna produto vao virar combustivel
data.head()

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,COMBUSTIVEL,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,COMBUSTIVEL,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,COMBUSTIVEL,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,COMBUSTIVEL,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,COMBUSTIVEL,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [145]:
produto_view

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004


In [146]:
produto_copy

,PRODUTO
0,ETANOL HIDRATADO
1,ETANOL HIDRATADO
2,ETANOL HIDRATADO
3,ETANOL HIDRATADO
4,ETANOL HIDRATADO
...,...
106818,GNV
106819,GNV
106820,GNV
106821,GNV


#### 1.3.4.2 Atribuindo listas ou series

In [147]:
nwros, ncols = data.shape
nwros, ncols

(106823, 21)

In [148]:
novos_produtos =[f'produto {i}'for i in range(nwros)]
len(novos_produtos)

106823

In [149]:
data['PRODUTO'] = novos_produtos
data

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,produto 0,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,produto 1,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,produto 2,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,produto 3,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,produto 4,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,produto 106818,7,R$/m3,3.574,0.065,...,3.690,0.818,0.018,2.756,0,2.756,2.756,0,6,2019
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,produto 106819,23,R$/m3,3.401,0.129,...,3.789,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,produto 106820,24,R$/m3,2.912,0.190,...,3.499,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,produto 106821,52,R$/m3,3.020,0.229,...,3.490,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019


In [150]:
data['PRODUTO'] = produto_copy
data

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,PREÇO MÁXIMO REVENDA,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,1.350,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,1.449,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,1.760,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,1.509,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,1.400,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,3.690,0.818,0.018,2.756,0,2.756,2.756,0,6,2019
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,3.789,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,3.499,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,3.490,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019


#### 1.3.4.3 Criando novas colunas
Para **criar uma nova coluna** em um data frame, basta atribuirmos uma _lista/Series de valores_ a uma **nova 'chave'** do data frame. <br/>

**PS:** A _quantidade de valores_ da lista precisa ser **igual** ao _número de linhas/registros_ do data frame.

In [151]:
data['coluna sem noçao'] = "default"
data

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,MARGEM MÉDIA REVENDA,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem noçao
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,0.463,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004,default
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,0.399,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004,default
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,0.419,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004,default
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,0.432,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004,default
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,0.24,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004,default
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,0.818,0.018,2.756,0,2.756,2.756,0,6,2019,default
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,0.95,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019,default
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,0.914,0.065,1.998,0,1.9981,1.9981,0,6,2019,default
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,0.646,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019,default


In [152]:
data['coluna a partir de lista'] = range(data.shape[0])
data

,Unnamed: 0,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,...,COEF DE VARIAÇÃO REVENDA,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem noçao,coluna a partir de lista
0,0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,...,0.012,0.825,0.11,0.4201,0.9666,0.133,5,2004,default,0
1,1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,...,0.098,0.763,0.088,0.5013,1.05,0.115,5,2004,default,1
2,2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,...,0.070,0.97,0.095,0.5614,1.161,0.098,5,2004,default,2
3,3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,...,0.055,0.83,0.119,0.5991,1.22242,0.143,5,2004,default,3
4,4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,...,0.066,0.941,0.077,0.7441,1.0317,0.082,5,2004,default,4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,...,0.018,2.756,0,2.756,2.756,0,6,2019,default,106818
106819,106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,...,0.038,2.451,0.402,1.9842,2.8661,0.164,6,2019,default,106819
106820,106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,...,0.065,1.998,0,1.9981,1.9981,0,6,2019,default,106820
106821,106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,...,0.076,2.374,0.165,2.0179,2.5093,0.07,6,2019,default,106821


In [153]:
data['nao funcinoa']= [1,2,3]
data

ValueError: Length of values (3) does not match length of index (106823)

<br/>

Outro exemplo:

In [ ]:
data.head()

In [ ]:
data["PREÇO DE MEDIA (dolares)"] = data["PREÇO MÉDIO REVENDA"] * 6.0
data

**PS:** Obviamente, a lógica correta em converter o preço dos combustíveis em reais para dólares não é considerar uma taxa de câmbio fixa, uma vez que cada preço foi aferido em um momento diferente.

### 1.3.4 Índices

Todo Data Frame possui **índices**, que não são considerado colunas da tabela. Tais índices são comumente **númericos**, de 0 a num_linhas-1, mas também podem ser **textuais (rótulos/labels)**.

In [ ]:
data

In [ ]:
data.index


Use `list(data.index)` ou `data.index.to_list()` para converter um RangeIndex para uma python list.

#### **Exemplo de Data Frame com Índices Textuais (labels)**

In [ ]:
pesquisa_de_satisfacao = pd.DataFrame({
    'bom': [50, 21, 100],
    'ruim': [131, 2, 30],
    'pessimo': [30, 20, 1]
}, index=['XboxOne', 'Playstation4', 'Switch'])

In [ ]:
pesquisa_de_satisfacao.head()

In [ ]:
pesquisa_de_satisfacao.index

In [ ]:
data.head()

### 1.4 Selecionando uma ou mais amostras (Indexação)

#### **==>  Index-based selection (seleção baseada em Índices)**
Mostrando linhas específicas de um DataFrame:

`iloc`: seleciona elementos do Dataframe, baseado em seu **índice (número)** --> row-first, column-second

**Selecionando uma amostra/linha:**

In [ ]:
data.iloc[1]

In [ ]:
data.iloc[0:6]

**Selecionando múltiplias amostras/linhas:**

In [ ]:
data.iloc[[5,9,10,15]]

In [ ]:
data.iloc[[5,2,11,7]]

In [ ]:
data.iloc[1,5]

In [ ]:
# e assim por diante!

#### **==>  Label-based selection (seleção baseadas em Rótulos)**

`loc`: seleciona elementos do Dataframe, baseado em seus **rótulos** --> row-first, column-second

In [ ]:
pesquisa_de_satisfacao

In [ ]:
pesquisa_de_satisfacao.iloc[0]


In [ ]:
pesquisa_de_satisfacao.loc["Switch"]

In [ ]:
pesquisa_de_satisfacao

In [ ]:
pesquisa_de_satisfacao.loc['Playstation4','ruim']

In [ ]:
pesquisa_de_satisfacao.loc[['XboxOne','Playstation4']]

In [ ]:
pesquisa_de_satisfacao[['bom','pessimo']]

### 1.5 Selecionando um ou mais atributos (colunas)

In [ ]:
data.head()

In [ ]:
data.ESTADO

In [ ]:
data.loc[:,'ESTADO']

In [ ]:
data[['ESTADO','PRODUTO','REGIÃO']]

Como o rótulo da coluna 'DATA INICIAL' **possui espaço**, não é possível acessá-la como método: `data.DATA INICIAL`

### 1.6 Removendo um Atributo (Coluna) do Data Frame

Como vimos anteriormente, o atributo 'Unnamed: 0' parece ser um **ruído** em nosso dataset. Desta maneira, vamos eliminá-lo,

In [ ]:
data.head()

In [ ]:
del data["Unnamed: 0"]

In [ ]:
data

### 1.7 Salvando um Data Frame

Para salvarmos um Data Frame para um **arquivo CSV**, basta usarmos o método `.to_csv`. <br/>
Por padrão, esse método **salva os índices da tabela como uma coluna no CSV**.<br/>
Como no geral tais índices são números de 0 a n-1, não há necessidade para isso (veja que removemos anteriormente a coluna 'Unnamed: 0' que foi justamente esse caso).<br/>
Desta forma, utilize o parâmetro: `index=False`.

Por padrão, o método utilizará a ',' como separador das colunas. Caso queira alterá-lo, utilize o parâmetro `sep`.

In [ ]:
data.to_csv( '/content/drive/MyDrive/trilia-dados/GasPricesinBrazil_2004-2019-copia.csv', index=False)

### 1.8 Seleção Condicional: Filtrando amostras

Durante nossas análise exploratórias, frequentemente filtraremos nossas amostras, a partir de certas **condições**, para fins de análise mais específica. <br/>
Existem algumas maneiras de fazermos tal filtragem. Antes disso, vamos carregar nosso dataset pré-processado que salvamos no item anterior.

In [ ]:
data = pd.read_csv('/content/drive/MyDrive/trilia-dados/GasPricesinBrazil_2004-2019-copia.csv')

In [ ]:
data.head()

In [ ]:
data["ESTADO"].unique()

#### **Selecionando apenas os preços dos Postos de São Paulo**

##### **==> Alternativa 1: Seleção Condicional (Comparações diretas)**

O código abaixo retorna uma ***Series ('array') de booleans***, com o número de linhas (amostras) do Data Frame, que informa os registros de preços dos postos do _estado de São Paulo_ (True).

In [ ]:
data["ESTADO"]== "SAO PAULO"

In [ ]:
selecao = data["ESTADO"]== "SAO PAULO"

In [ ]:
selecao

In [ ]:
type(selecao)

In [ ]:
selecao.shape

In [ ]:
data.shape

Para **filtrarmos** os registros de postos do estado de São Paulo:

In [ ]:
data[selecao]

O resultado é um Data Frame com _apenas_ os registros desejados após a **filtragem**.<br/>
Podemos ainda utilizar o método `loc` para o mesmo fim:

##### **==> Alternativa 2: Utilizando o método `query`**

`query` filtra linhas de um DataFrame baseado em uma **query (pergunta)**.

In [ ]:
data.head()


In [ ]:
data.query('ESTADO =="SAO PAULO"')

<br/>

Uma boa prática é **salvar o Data Frame filtrado em uma nova variável**. Isso simplifica a complexidade do código para futuras análise feita para os postos de São Paulo.

In [ ]:
posto_sp = data.query('ESTADO =="SAO PAULO"')

In [ ]:
posto_sp

In [ ]:
type(posto_sp)

In [ ]:
posto_sp.shape

In [ ]:
posto_sp.head()

In [ ]:
posto_sp.reset_index()

In [ ]:
posto_sp.reset_index(drop=True)

In [ ]:
posto_sp.reset_index(drop=True,inplace=True)

In [ ]:
posto_sp

In [ ]:
posto_sp = data.query('ESTADO=="SAO PAULO"').reset_index(drop=True)

In [ ]:
posto_sp

#### **Selecionando registros de postos do Rio de Janeiro com Preços acima de 2 reais**

In [ ]:
data.head()

In [ ]:
data['ESTADO'] == 'RIO JANIERO'

In [ ]:
data["PREÇO MÉDIO REVENDA"] > 2

In [ ]:
SELECAO = (data['ESTADO'] == 'RIO JANEIRO') & data["PREÇO MÉDIO REVENDA"] > 2
SELECAO

Note que o resultado da seleção continua sendo uma _Series de booleans_ com o _mesmo número de linhas/amostras do DataFrame_, de modo que cada linha possuirá um valor booleano indicando se o posto é do Rio de Janeiro e o preço aferido do combustível é maior que 2 reais (True) ou não.

O símbolo **&** significa **AND** na comparação. Essa nomenclatura do python/pandas é diferente das nomenclaturas tradicionais (&&). <br/>
Similarmente:
- **|** representa o **OR** (não é ||)
- **~** representa o **NOT** (não é !)

In [ ]:
data[SELECAO]

Alternativamente, poderíamos usar o método `query` para fazermos tal seleção. Porém, isso não é possível especificamente para esse caso, pois o rótulo da coluna 'PREÇO MÉDIO REVENDA' possui caracteres inválidos para o método (cedilha, acentos)

In [ ]:
# Não funciona
# data.query('ESTADO=="RIO DE JANEIRO" and PREÇO MÉDIO REVENDA > 2')

In [ ]:
data.query('ESTADO=="RIO DE JANEIRO" or ESTADO == "SAO PAULO"')

**Aprofundando mais ainda**

A primeira comparação `(data['ESTADO'] == 'RIO DE JANEIRO')` checa, linha a linha (amostra a amostra) do DataFrame, quais são aquelas cujo o estado é RIO DE JANEIRO. Nenhuma averiguação de preços é feita nesse momento. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

A segunda comparação `(data['PREÇO MÉDIO REVENDA'] > 2)` checa, linha a linha (amostra a amostra) do DataFrame, quais são os registro cujo preço do combustível é maior que 2 reais. Note que essa comparação checará os postos de **TODOS os estados**. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

Por fim, as duas "perguntas" são unidas pelo AND (&) que retorna a "pergunta completa" que fizemos.

Alguns podem argumentar que tal abordagem é **ineficiente**, uma vez que, para cada condição ("pergunta"), estamos varrendo todas as linhas do DataFrame. <br/>
O Pandas _tenta otimizar_ isso ao máximo por de trás dos panos. Mas, de fato, de tivermos um dataset **muito grande** (centenas de milhares de linhas), tal abordagem se tornará _lenta_.

Assim, poderíamos fazer filtragem com múltiplos condicionais em partes:

In [ ]:
seleçao1 =(data['ESTADO'] == 'RIO DE JANEIRO')
posto_rj = data[seleçao1]
posto_rj

In [ ]:
seleçao2 = (posto_rj['PREÇO MÉDIO REVENDA'] > 2)
seleçao2

In [ ]:
preço_maior_que_2 =posto_rj[seleçao2]
preço_maior_que_2

#### **Selecionando registros de postos de São Paulo ou do Rio de Janeiro com Gasolina Comum acima de 2 reais**

Podemos fazer a solução do "jeito mais lento", percorrendo o DataFrame inteiro _múltiplas vezes_:

In [ ]:
gasolina_comum = (data['ESTADO'] == "SAO PAULO")
gasolina_comum2 =(data['PRODUTO'] == 'GASOLINA COMUM')
gasolina_comum3 = (data['PREÇO MÉDIO REVENDA']> 2)
gasolina_comum_final = (gasolina_comum & gasolina_comum2 & gasolina_comum3)
gasolina_comum_final

In [ ]:
data[gasolina_comum_final]

<br/>

Alternativamente:

#### **Selecionando registros dos anos de 2008, 2010 e 2012**

**ALTERNATIVA 1**

In [ ]:
anos = (data['ANO'] == 2008) | (data['ANO'] == 2010) | (data['ANO'] == 2012)
data[anos]

**ALTERNATIVA 2**

In [ ]:
lista_de_anos =(2008,2010,2012)
data['ANO'].isin(lista_de_anos)


### **Iterando com DataFrames**

#### For-each `DataFrame.iterrows()` (LENTO ==> apenas indicado para iterar pequenos conjunto de dados)

In [ ]:
for index ,row in data.head(10).iterrows():
  print(f"indice {index} ==> {row['ESTADO']}")

<h2>2. Preparação dos dados</h2>
<hr/>

### 2.1 Removendo amostras com valores vazios (null / nan) no dataset

In [ ]:
data.info()

De um total de 106823 observações, **não há valores null** / nan para nenhum atributo. Mas, veremos que não é bem assim neste caso específico.<br/><br/>

### 2.2 Conversão de tipos de atributos

O pandas automaticamente reconhece os tipos de dados de cada coluna. <br/>
Porém, existem alguns atributos que estão com seus tipos errados: P. ex., "PREÇO MÉDIO DISTRIBUIÇÃO" deveria ser ```float64``` e não ```object```.<br/>
Nestes casos, muito provavelmente algumas amostras têm um string ao invés de um número para tais atributos. <br/>

Os atributos *"DATA INICIAL"* e *"DATA FINAL"* deveriam ser do tipo `datetime`.

Em outros casos, alguns **atributos categóricos** são ```objects```, mas poderiam ter o tipo ```category```, que é um tipo especial do pandas. <br/>
Este tipo é necessário para se utilizar algumas funções específicas do pandas. <br/>
**Não** converteremos para este tipo por ora.

In [ ]:
data_pre = data.copy()

#### **Datas**
Como os atributos de data do datset já estão em um formato de data aceitável (YYYY-MM-DD), não precisamos forçar nenhuma conversão nesse sentido.

In [ ]:
data_pre['DATA INICIAL']= pd.to_datetime(data_pre['DATA INICIAL'])
data_pre['DATA FINAL']= pd.to_datetime(data_pre['DATA FINAL'])


In [ ]:
data_pre.info()

#### **Dados Numéricos**

In [ ]:
for atributo in ['MARGEM MÉDIA REVENDA','PREÇO MÉDIO DISTRIBUIÇÃO', 'DESVIO PADRÃO DISTRIBUIÇÃO','PREÇO MÍNIMO DISTRIBUIÇÃO','PREÇO MÁXIMO DISTRIBUIÇÃO']:
 data_pre[atributo]= pd.to_numeric(data_pre[atributo],errors = 'coerce')


In [ ]:
data_pre.info()

<br/>

Note que temos vários valores ***null*** agora **após a *conversão de tipos***. Vamos checá-los com mais cuidado nos dados originais e preprocessados.

### 2.3 Limpeza de dados

In [ ]:
mask = data_pre['PREÇO MÉDIO DISTRIBUIÇÃO'].isnull()

In [ ]:
data_pre[mask]

In [ ]:
data[mask]

Várias amostras possuem a _string '-'_ em algumas colunas ao invés de um número de fato. Ou seja, não há aferições destes atributos para estas amostras. <br/>

<br/>

Poderíamos **preencher os valores NaN com um valor padrão**. Para isso, basta usar o método `.fillna`.

In [ ]:
data_pre_fill = data_pre.fillna(0)
data_pre_fill

In [ ]:
data_pre_fill[mask]

In [ ]:
data_pre[mask]

In [ ]:
data_pre_fill = data_pre.fillna(value={ 'PREÇO MÉDIO DISTRIBUIÇÃO':10,
                                          'DESVIO PADRÃO DISTRIBUIÇÃO':20,
                                           'PREÇO MIŃIMO DISTRIBUIÇÃO': 30,
                                           'PREÇO MÁXIMO DISTRIBUIÇÃO': 'vazio'})


In [ ]:
data_pre_fill[mask]

<br/>

Por mais que a função `fillna` seja interessante e útil em muitos casos, no problema em questão estamos interessados em analisar precisamente, p. ex., o **'PREÇO MÉDIO DISTRIBUIÇÃO'**.<br/>
A fim de não termos valores _sintéticos_ gerados pelo `fillna`, que possam atrapalhar nossa análise, iremos **remover (drop) todas as amostras que possuem qualquer valor NaN** para quaisquer atributos/colunas. <br/>

Para isso, basta utilizarmos o método `dropna`.

In [ ]:
data_pre.dropna(inplace=True)

In [ ]:
data_pre.info()

Nosso data frame agora, após essa _limpeza_, ficou mais enxuto, contentdo 103392 registros frente aos 106823 registros originais. <br/>

Essas são apenas algumas das possíveis _técnicas de limpeza de dados_. Outras estratégias, p. ex., **confiam na detecção de outliers**, que não veremos neste curso.

#### **Salvando o Dataset Preprocessado**

In [ ]:
data_pre.to_csv('/content/drive/MyDrive/trilia-dados/GasPricesinBrazil_2004-2019-prepocessando_final.csv')

<h2>3. Estatísticas Descritivas</h2>
<hr/>

O Pandas fornecem algumas funções/métodos ue computam certas estatísticas descritivas.

`describe`: exibe várias **estatísticas descritivas** para os _atributos_ de um dataframe ou para uma _Series_.

In [156]:
data_final = pd.read_csv('/content/drive/MyDrive/trilia-dados/GasPricesinBrazil_2004-2019-copia.csv')


In [ ]:
data_final.describe()

In [ ]:
coluna_removida = data_final.pop('coluna sem noçao')
coluna_removida = data_final.pop ('coluna a partir de lista')
coluna_removida = data_final.pop ('PREÇO DE MEDIA (dolares)')


<br/>

Como o resultado do `describe` de um _dataframe_ é outro _dataframe_, podemos filtrar apenas algumas colunas.

In [ ]:
data_final['PREÇO MÉDIO REVENDA'].describe()

In [ ]:
data_final.describe()['PREÇO MÉDIO REVENDA']

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
stats = data_final.describe()
stats

In [ ]:
stats[['PREÇO MÉDIO REVENDA','PREÇO MEÉDIO DISTRIBUIÇÃO']]

**Acessando apenas algumas estatísticas**

In [ ]:
stats

In [ ]:
stats.loc[['min','max','mean'],'PREÇO MÉDIO REVENDA']

In [ ]:
stats.loc[['min','max','mean'],['PREÇO MÉDIO REVENDA','PREÇO MÉDIO DISTRIBUIÇÃO']]

<br/>

`mean`, `std`, `min`, etc: cada uma das estatísticas do `describe` podem ser computadas individualmente:

In [ ]:
data_final.head()

#### Qual é o menor preço mínimo de revenda?

In [ ]:
data_final['PREÇO MÍNIMO REVENDA'].min()

#### Qual é a média e desvio padrão dos preços mínimos de revenda?

In [ ]:
data_final['PREÇO MÍNIMO REVENDA'].mean()

#### Quais são os estados considerados?

In [ ]:
data_final['ESTADO']

In [ ]:
data_final['ESTADO'].unique()

#### Quantos registros (aferições) cada estado possui?

`.value_counts()`:  Conta a frequência dos valores de uma dada variável (de preferência, _categórica_).

In [ ]:
data_final['ESTADO'].value_counts()

In [ ]:

data_final['ESTADO'].value_counts().to_frame()

<h2>4. Executando funções para cada item de um DataFrame ou Series</h2>
<hr/>

Uma alternativa ao `for-loop` que vimos anteriormente e que é _lento_, é usarmos _funções próprias do pandas_ que **aplicam/mapeiam uma dada função a todos os elementos de um DataFrame ou Series**, retornando novos elementos "transformados".


<img src='./imagens/apply_map_applymap.png' width=300/>


Fonte: https://towardsdatascience.com/introduction-to-pandas-apply-applymap-and-map-5d3e044e93ff

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4],
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]},
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])

In [ ]:
df

`apply()`: usado para aplicar uma função ao longo de um eixo de um DataFrame ou em valores de uma Series.

<img src='./imagens/pandas_axis.jpg' width=500/>

Fonte: https://www.allthesnippets.com/browse/pandas/df_axis.html

In [ ]:
def nossa_soma(linha):
  return linha.sum()

In [ ]:
df['SOMA(A,B,C)'] = df.apply(nossa_soma,axis=1)
df

<img src='./imagens/apply_axis_1.png' width=250/>

In [ ]:
df.apply(nossa_soma,axis=0)
df

<img src='./imagens/apply_axis_0.png' width=250/>

##### Usando `lambda` functions

In [ ]:
df['MEDIA(A,B,C)']= df[['A','B','C']].apply(lambda series:series.mean(),axis=1)
df

<img src='./imagens/apply_axis_1_mean.png' width=350/>

In [ ]:
df['C*2'] = df['C'].apply(lambda x: x * 2)
df

<br/>

`applymap()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de um DataFrame.

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4],
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]},
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])
df

In [ ]:
df.applymap(lambda x: x **2)

In [ ]:
df

<br/>

`map()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de uma _Series_.

In [ ]:
nomes = pd.Series(['Joao', 'Maria', 'Alice' , 'Pedro'])
nomes

In [ ]:
nomes.map(lambda x: x.upper())

In [ ]:
nomes.str.upper()

<h2>5. Agrupamento</h2>
<hr/>

`groupby`: Usado para criar **grupo de elementos** (e.x., baseado nos valores de um atributo). <br/>
**Funções** podem então ser aplicadas para os _elementos de cada grupo_, de modo que os **resultados de cada grupo são combinados**.

In [157]:
data_final.head()

,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,...,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem noçao,coluna a partir de lista,PREÇO DE MEDIA (dolares)
0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,1.19,...,0.825,0.11,0.4201,0.9666,0.133,5,2004,default,0,7.728
1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,0.89,...,0.763,0.088,0.5013,1.05,0.115,5,2004,default,1,6.972
2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,1.18,...,0.97,0.095,0.5614,1.161,0.098,5,2004,default,2,8.334
3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,1.09,...,0.83,0.119,0.5991,1.22242,0.143,5,2004,default,3,7.572
4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,1.05,...,0.941,0.077,0.7441,1.0317,0.082,5,2004,default,4,7.086


In [161]:
grupos = data_final.groupby('REGIÃO')
grupos

In [162]:
grupos.groups

{'CENTRO OESTE': [0, 1, 2, 3, 27, 28, 29, 30, 54, 55, 56, 57, 81, 82, 83, 84, 108, 109, 110, 111, 135, 136, 137, 138, 162, 163, 164, 165, 189, 190, 191, 192, 216, 217, 218, 219, 243, 244, 245, 246, 270, 271, 272, 273, 297, 298, 299, 300, 324, 325, 326, 327, 351, 352, 353, 354, 378, 379, 380, 381, 405, 406, 407, 408, 432, 433, 434, 435, 459, 460, 461, 462, 486, 487, 488, 489, 513, 514, 515, 516, 540, 541, 542, 543, 567, 568, 569, 570, 594, 595, 596, 597, 621, 622, 623, 624, 648, 649, 650, 651, ...], 'NORDESTE': [4, 5, 6, 7, 8, 9, 10, 11, 12, 31, 32, 33, 34, 35, 36, 37, 38, 39, 58, 59, 60, 61, 62, 63, 64, 65, 66, 85, 86, 87, 88, 89, 90, 91, 92, 93, 112, 113, 114, 115, 116, 117, 118, 119, 120, 139, 140, 141, 142, 143, 144, 145, 146, 147, 166, 167, 168, 169, 170, 171, 172, 173, 174, 193, 194, 195, 196, 197, 198, 199, 200, 201, 220, 221, 222, 223, 224, 225, 226, 227, 228, 247, 248, 249, 250, 251, 252, 253, 254, 255, 274, 275, 276, 277, 278, 279, 280, 281, 282, 301, ...], 'NORTE': [13, 14, 1

In [163]:
grupos.indices

{'CENTRO OESTE': array([     0,      1,      2, ..., 106791, 106792, 106812]),
 'NORDESTE': array([     4,      5,      6, ..., 106816, 106818, 106822]),
 'NORTE': array([    13,     14,     15, ..., 106802, 106803, 106807]),
 'SUDESTE': array([    20,     21,     22, ..., 106813, 106817, 106821]),
 'SUL': array([    24,     25,     26, ..., 106815, 106819, 106820])}

In [164]:
grupos.get_group

<bound method BaseGroupBy.get_group of <pandas.core.groupby.generic.DataFrameGroupBy object at 0x7cb52da36ae0>>

In [165]:
grupos.describe()

NÚMERO DE POSTOS PESQUISADOS                                      \
                                    count        mean         std  min    25%   
REGIÃO                                                                          
CENTRO OESTE                      15429.0  137.263983  107.396548  1.0   76.0   
NORDESTE                          36869.0  123.194364  112.856327  1.0   40.0   
NORTE                             24702.0   66.249899   53.119277  1.0   29.0   
SUDESTE                           17047.0  713.297002  786.310843  1.0  158.0   
SUL                               12776.0  353.486146  238.387934  1.0  153.0   

                                   PREÇO MÉDIO REVENDA             ...  \
                50%    75%     max               count       mean  ...   
REGIÃO                                                             ...   
CENTRO OESTE  119.0  178.0   766.0             15429.0  11.843486  ...   
NORDESTE       93.0  167.0   625.0             36869.0  10.097994  ...   
NORTE          52.0   94.0   383.0             24702.0  12.670958  ...   
SUDESTE       403.0  993.0  4167.0             17047.0   9.700178  ...   
SUL           362.0  560.0  1416.0             12776.0  10.007542  ...   

             coluna a partir de lista           PREÇO DE MEDIA (dolares)  \
                                  75%       max                    count   
REGIÃO                                                                     
CENTRO OESTE                 80018.00  106812.0                  15429.0   
NORDESTE                     80098.00  106822.0                  36869.0   
NORTE                        80299.50  106807.0                  24702.0   
SUDESTE                      80048.50  106821.0                  17047.0   
SUL                          80059.75  106820.0                  12776.0   

                                                                              
                   mean         std    min     25%     50%      75%      max  
REGIÃO                                                                        
CENTRO OESTE  71.060919  117.624654  6.582  12.606  16.518  22.6560  596.142  
NORDESTE      60.587964   98.749681  5.850  11.928  15.858  21.8460  477.780  
NORTE         76.025748  118.116441  7.314  14.202  17.958  25.3800  523.620  
SUDESTE       58.201066   97.031261  4.596  11.274  15.276  21.2850  475.152  
SUL           60.045250  100.287489  5.532  11.862  15.459  21.0675  432.528  

[5 rows x 80 columns]

In [168]:
grupos.mean()

TypeError: agg function failed [how->mean,dtype->object]

In [169]:
grupos.min()

,DATA INICIAL,DATA FINAL,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,PREÇO MÁXIMO REVENDA,...,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem noçao,coluna a partir de lista,PREÇO DE MEDIA (dolares)
REGIÃO,,,,,,,,,,,,,,,,,,,,,
CENTRO OESTE,2004-05-09,2004-05-15,DISTRITO FEDERAL,ETANOL HIDRATADO,1,R$/13Kg,1.097,0.0,0.87,1.099,...,-,-,-,-,-,1,2004,default,0,6.582
NORDESTE,2004-05-09,2004-05-15,ALAGOAS,ETANOL HIDRATADO,1,R$/13Kg,0.975,0.0,0.79,0.999,...,-,-,-,-,-,1,2004,default,4,5.850
NORTE,2004-05-09,2004-05-15,ACRE,ETANOL HIDRATADO,1,R$/13Kg,1.219,0.0,1.05,1.219,...,-,-,-,-,-,1,2004,default,13,7.314
SUDESTE,2004-05-09,2004-05-15,ESPIRITO SANTO,ETANOL HIDRATADO,1,R$/13Kg,0.766,0.0,0.59,1.099,...,-,-,-,-,-,1,2004,default,20,4.596
SUL,2004-05-09,2004-05-15,PARANA,ETANOL HIDRATADO,1,R$/13Kg,0.922,0.0,0.73,1.199,...,-,-,-,-,-,1,2004,default,24,5.532


In [170]:
data_final.groupby('REGIÃO').min

<bound method GroupBy.min of <pandas.core.groupby.generic.DataFrameGroupBy object at 0x7cb52daa4740>>

<br/>

Também podemos ter agrupamentos por mais de um atributo.

In [171]:
data_final.head()

,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,...,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem noçao,coluna a partir de lista,PREÇO DE MEDIA (dolares)
0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,1.19,...,0.825,0.11,0.4201,0.9666,0.133,5,2004,default,0,7.728
1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,0.89,...,0.763,0.088,0.5013,1.05,0.115,5,2004,default,1,6.972
2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,1.18,...,0.97,0.095,0.5614,1.161,0.098,5,2004,default,2,8.334
3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,1.09,...,0.83,0.119,0.5991,1.22242,0.143,5,2004,default,3,7.572
4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,1.05,...,0.941,0.077,0.7441,1.0317,0.082,5,2004,default,4,7.086


Qual e o  preço médio  de cada  Produto (COMBUSTIVIES) pra cada **regiao**

In [172]:
GRUPOS = data_final.groupby(['REGIÃO','PRODUTO'])
GRUPOS

In [173]:
GRUPOS.groups

{('CENTRO OESTE', 'ETANOL HIDRATADO'): [0, 1, 2, 3, 27, 28, 29, 30, 54, 55, 56, 57, 81, 82, 83, 84, 108, 109, 110, 111, 135, 136, 137, 138, 162, 163, 164, 165, 189, 190, 191, 192, 216, 217, 218, 219, 243, 244, 245, 246, 270, 271, 272, 273, 297, 298, 299, 300, 324, 325, 326, 327, 351, 352, 353, 354, 378, 379, 380, 381, 405, 406, 407, 408, 432, 433, 434, 435, 459, 460, 461, 462, 486, 487, 488, 489, 513, 514, 515, 516, 540, 541, 542, 543, 567, 568, 569, 570, 594, 595, 596, 597, 621, 622, 623, 624, 648, 649, 650, 651, ...], ('CENTRO OESTE', 'GASOLINA COMUM'): [12064, 12065, 12066, 12067, 12091, 12092, 12093, 12094, 12118, 12119, 12120, 12121, 12145, 12146, 12147, 12148, 12172, 12173, 12174, 12175, 12199, 12200, 12201, 12202, 12226, 12227, 12228, 12229, 12253, 12254, 12255, 12256, 12280, 12281, 12282, 12283, 12307, 12308, 12309, 12310, 12334, 12335, 12336, 12337, 12361, 12362, 12363, 12364, 12388, 12389, 12390, 12391, 12415, 12416, 12417, 12418, 12442, 12443, 12444, 12445, 12469, 12470, 124

In [174]:
GRUPOS.mean

TypeError: agg function failed [how->mean,dtype->object]

In [178]:
GRUPOS['PREÇO MÉDIO REVENDA'].describe()

count       mean        std     min       25%  \
REGIÃO       PRODUTO                                                            
CENTRO OESTE ETANOL HIDRATADO  3140.0   2.135950   0.561154   1.162   1.71600   
             GASOLINA COMUM    3140.0   3.078123   0.616315   1.936   2.66800   
             GLP               3137.0  48.181149  15.189152  28.791  37.34000   
             GNV               1523.0   2.004153   0.457802   1.097   1.59900   
             ÓLEO DIESEL       3140.0   2.481534   0.627607   1.435   2.03300   
             ÓLEO DIESEL S10   1349.0   3.241185   0.433797   2.212   2.82000   
NORDESTE     ETANOL HIDRATADO  7064.0   2.322075   0.605121   1.181   1.82500   
             GASOLINA COMUM    7065.0   3.002337   0.641163   1.949   2.58400   
             GLP               7061.0  42.221312  11.654978  28.367  33.59800   
             GNV               5575.0   1.925318   0.555183   0.975   1.60300   
             ÓLEO DIESEL       7065.0   2.347550   0.606381   1.313   1.90000   
             ÓLEO DIESEL S10   3039.0   3.041957   0.442555   2.130   2.60000   
NORTE        ETANOL HIDRATADO  5404.0   2.530796   0.655421   1.226   2.04600   
             GASOLINA COMUM    5494.0   3.149212   0.640964   1.995   2.70525   
             GLP               5494.0  47.174539  14.477296  28.086  36.50925   
             GNV                457.0   2.000190   0.602366   1.219   1.39900   
             ÓLEO DIESEL       5494.0   2.569905   0.650350   1.450   2.08800   
             ÓLEO DIESEL S10   2359.0   3.310619   0.533966   2.172   2.87800   
SUDESTE      ETANOL HIDRATADO  3140.0   2.164832   0.652355   0.766   1.69275   
             GASOLINA COMUM    3140.0   3.000565   0.698573   1.888   2.50200   
             GLP               3139.0  42.077813  11.522013  28.026  33.63850   
             GNV               3136.0   1.779684   0.507284   0.996   1.47300   
             ÓLEO DIESEL       3140.0   2.339699   0.588641   1.356   1.87600   
             ÓLEO DIESEL S10   1352.0   3.054425   0.419618   2.180   2.64700   
SUL          ETANOL HIDRATADO  2354.0   2.268864   0.703567   0.922   1.72100   
             GASOLINA COMUM    2355.0   2.973136   0.624032   1.961   2.54300   
             GLP               2355.0  43.535465  11.628193  29.440  34.82450   
             GNV               2343.0   1.907944   0.500694   1.168   1.51200   
             ÓLEO DIESEL       2355.0   2.332862   0.528454   1.374   1.95100   
             ÓLEO DIESEL S10   1014.0   2.981745   0.366952   2.140   2.66300   

                                   50%       75%     max  
REGIÃO       PRODUTO                                      
CENTRO OESTE ETANOL HIDRATADO   2.0010   2.47900   3.999  
             GASOLINA COMUM     2.8560   3.46025   4.963  
             GLP               42.8570  55.90500  99.357  
             GNV                1.9550   2.35250   3.890  
             ÓLEO DIESEL        2.2455   3.01900   4.162  
             ÓLEO DIESEL S10    3.2640   3.58400   4.200  
NORDESTE     ETANOL HIDRATADO   2.1880   2.67550   3.927  
             GASOLINA COMUM     2.7420   3.42500   4.951  
             GLP               37.5750  48.18300  79.630  
             GNV                1.8210   2.16850   3.811  
             ÓLEO DIESEL        2.0820   2.87300   4.114  
             ÓLEO DIESEL S10    3.0870   3.33800   4.069  
NORTE        ETANOL HIDRATADO   2.3430   2.93525   4.459  
             GASOLINA COMUM     2.9320   3.57975   5.362  
             GLP               42.4535  54.63575  87.270  
             GNV                1.8500   2.30000   4.020  
             ÓLEO DIESEL        2.3340   3.08000   4.774  
             ÓLEO DIESEL S10    3.3070   3.65300   4.878  
SUDESTE      ETANOL HIDRATADO   2.0435   2.57200   4.115  
             GASOLINA COMUM     2.7930   3.45150   5.119  
             GLP               38.2660  49.25100  79.192  
             GNV                1.6940   2.00300   3.594  
             ÓLEO DIESE

<br/>

`.agg`: **agrega (roda)** uma série de funções para os elementos de um dataframe ou de grupos de um dataframe.

In [180]:
df = pd.DataFrame([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9],
                   [None,None,None]],
                  columns=['A', 'B', 'C'])
df

,A,B,C
0,1.0,2.0,3.0
1,4.0,5.0,6.0
2,7.0,8.0,9.0
3,NaN,NaN,NaN


In [181]:
df.agg([sum,min])

/tmp/ipython-input-1413054331.py:1: FutureWarning: The provided callable <built-in function sum> is currently using Series.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  df.agg([sum,min])
/tmp/ipython-input-1413054331.py:1: FutureWarning: The provided callable <built-in function min> is currently using Series.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  df.agg([sum,min])


,A,B,C
sum,12.0,15.0,18.0
min,1.0,2.0,3.0


In [182]:
grupos = data_final.groupby("REGIÃO")
grupos

In [184]:
grupos['PREÇO MÉDIO REVENDA'].agg([min,max])

/tmp/ipython-input-2084642471.py:1: FutureWarning: The provided callable <built-in function min> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  grupos['PREÇO MÉDIO REVENDA'].agg([min,max])
/tmp/ipython-input-2084642471.py:1: FutureWarning: The provided callable <built-in function max> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  grupos['PREÇO MÉDIO REVENDA'].agg([min,max])


,min,max
REGIÃO,,
CENTRO OESTE,1.097,99.357
NORDESTE,0.975,79.630
NORTE,1.219,87.270
SUDESTE,0.766,79.192
SUL,0.922,72.088


<h2>6. Ordenação</h2>
<hr/>

In [185]:
notas = pd.DataFrame({
    'nome': ['João', 'Maria', 'José', 'Alice'],
    'idade': [20, 21, 19, 20],
    'nota_final': [5.0, 10.0, 6.0, 10.0]
})
notas

,nome,idade,nota_final
0,João,20,5.0
1,Maria,21,10.0
2,José,19,6.0
3,Alice,20,10.0


`.sort_values()`: ordena valores ao longo de um eixo.

In [188]:
notas.sort_values(by='nota_final')

,nome,idade,nota_final
0,João,20,5.0
2,José,19,6.0
1,Maria,21,10.0
3,Alice,20,10.0


Por padrão, o método retorna uma cópia dos dados ordenados em **ordem crescente (ascendente)**. Podemos alterar isso pelo argumento `ascending`.

In [193]:
notas.sort_values(by=['nota_final','nome'],ascending=[False, True],inplace=True)

In [194]:
notas

,nome,idade,nota_final
3,Alice,20,10.0
1,Maria,21,10.0
2,José,19,6.0
0,João,20,5.0


<br/>

Podemos ordenar a partir de **mais de uma coluna**:

Ordena os registros, primeiramente, pela coluna 'nota_final' em **ordem descrente**. <br/>
Então, reordena os registros _"empatados"_, ou seja, com a **mesma nota final**, em _ordem alfabética_ (ordem crescente).

<h2>7. Exercícios</h2>
<hr/>

Vamos aplicar os conceitos que vimos em alguns exercícios. <br/>
Para isso, utilizaremos o dataset de _preços de combustíveis no Brasil_.

Como há apenas medições de janeiro a junho para o ano de 2019, resolvemos **remover os dados** deste ano da análise.

In [207]:
data_final.query('ANO != 2019')
default = data_final

Note que temos um novo dataframe com 99739 linhas, mas com índices fora desse intervalo. <br/>
Acontece que os registros mantiveram seus índices originais após a query. <br/>

Para resetar os índices de _0 a num_linhas-1_, basta usarmos o método `.reset_index()`.

Os índices agora foram **resetados**. Porém, os índices antigos se transformaram em _uma nova coluna_ chamada 'index'. <br/>
Para removê-la durante o _reset_, basta passarmos o argumento `drop=True`.

### 7.1 Qual a proporção de postos pesquisados para cada combustível em cada região

In [209]:
grupos = default.groupby("PRODUTO")
default

In [206]:
d

,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,...,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem noçao,coluna a partir de lista,PREÇO DE MEDIA (dolares)
0,2004-05-09,2004-05-15,CENTRO OESTE,DISTRITO FEDERAL,ETANOL HIDRATADO,127,R$/l,1.288,0.016,1.190,...,0.825,0.11,0.4201,0.9666,0.133,5,2004,default,0,7.728
1,2004-05-09,2004-05-15,CENTRO OESTE,GOIAS,ETANOL HIDRATADO,387,R$/l,1.162,0.114,0.890,...,0.763,0.088,0.5013,1.05,0.115,5,2004,default,1,6.972
2,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO,ETANOL HIDRATADO,192,R$/l,1.389,0.097,1.180,...,0.97,0.095,0.5614,1.161,0.098,5,2004,default,2,8.334
3,2004-05-09,2004-05-15,CENTRO OESTE,MATO GROSSO DO SUL,ETANOL HIDRATADO,162,R$/l,1.262,0.070,1.090,...,0.83,0.119,0.5991,1.22242,0.143,5,2004,default,3,7.572
4,2004-05-09,2004-05-15,NORDESTE,ALAGOAS,ETANOL HIDRATADO,103,R$/l,1.181,0.078,1.050,...,0.941,0.077,0.7441,1.0317,0.082,5,2004,default,4,7.086
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
106818,2019-06-23,2019-06-29,NORDESTE,RIO GRANDE DO NORTE,GNV,7,R$/m3,3.574,0.065,3.499,...,2.756,0,2.756,2.756,0,6,2019,default,106818,21.444
106819,2019-06-23,2019-06-29,SUL,RIO GRANDE DO SUL,GNV,23,R$/m3,3.401,0.129,3.230,...,2.451,0.402,1.9842,2.8661,0.164,6,2019,default,106819,20.406
106820,2019-06-23,2019-06-29,SUL,SANTA CATARINA,GNV,24,R$/m3,2.912,0.190,2.699,...,1.998,0,1.9981,1.9981,0,6,2019,default,106820,17.472
106821,2019-06-23,2019-06-29,SUDESTE,SAO PAULO,GNV,52,R$/m3,3.020,0.229,2.699,...,2.374,0.165,2.0179,2.5093,0.07,6,2019,default,106821,18.120


In [211]:
grupos['REGIÃO'].value_counts()

PRODUTO           REGIÃO      
ETANOL HIDRATADO  NORDESTE        7064
                  NORTE           5404
                  CENTRO OESTE    3140
                  SUDESTE         3140
                  SUL             2354
GASOLINA COMUM    NORDESTE        7065
                  NORTE           5494
                  CENTRO OESTE    3140
                  SUDESTE         3140
                  SUL             2355
GLP               NORDESTE        7061
                  NORTE           5494
                  SUDESTE         3139
                  CENTRO OESTE    3137
                  SUL             2355
GNV               NORDESTE        5575
                  SUDESTE         3136
                  SUL             2343
                  CENTRO OESTE    1523
                  NORTE            457
ÓLEO DIESEL       NORDESTE        7065
                  NORTE           5494
                  CENTRO OESTE    3140
                  SUDESTE         3140
                  SUL             2355
ÓLEO DIESEL S10   NORDESTE        3039
                  NORTE           2359
                  SUDESTE         1352
                  CENTRO OESTE    1349
                  SUL             1014
Name: count, dtype: int64

### 7.2 Como os preços da Gasolina Comum em São Paulo variaram em 2018?

In [214]:
gasolina_sp_2018 =default.query('PRODUTO == "GASOLINA COMUM" and ESTADO == "SAO PAULO" and ANO== 2018' )
gasolina_sp_2018.head()

,DATA INICIAL,DATA FINAL,REGIÃO,ESTADO,PRODUTO,NÚMERO DE POSTOS PESQUISADOS,UNIDADE DE MEDIDA,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,...,PREÇO MÉDIO DISTRIBUIÇÃO,DESVIO PADRÃO DISTRIBUIÇÃO,PREÇO MÍNIMO DISTRIBUIÇÃO,PREÇO MÁXIMO DISTRIBUIÇÃO,COEF DE VARIAÇÃO DISTRIBUIÇÃO,MÊS,ANO,coluna sem noçao,coluna a partir de lista,PREÇO DE MEDIA (dolares)
95351,2018-01-07,2018-01-13,SUDESTE,SAO PAULO,GASOLINA COMUM,1641,R$/l,3.988,0.156,3.499,...,3.577,0.104,2.99,3.951,0.029,1,2018,default,95351,23.928
95502,2018-01-14,2018-01-20,SUDESTE,SAO PAULO,GASOLINA COMUM,1648,R$/l,4.002,0.155,3.549,...,3.59,0.102,2.99,3.9014,0.028,1,2018,default,95502,24.012
95654,2018-01-21,2018-01-27,SUDESTE,SAO PAULO,GASOLINA COMUM,1648,R$/l,4.006,0.160,3.499,...,3.596,0.102,2.99,3.9014,0.028,1,2018,default,95654,24.036
95806,2018-01-28,2018-02-03,SUDESTE,SAO PAULO,GASOLINA COMUM,1644,R$/l,4.019,0.156,3.579,...,3.596,0.107,2.924,3.9537,0.03,1,2018,default,95806,24.114
95958,2018-02-04,2018-02-10,SUDESTE,SAO PAULO,GASOLINA COMUM,1646,R$/l,4.004,0.157,3.579,...,3.588,0.102,2.99,3.9537,0.028,2,2018,default,95958,24.024


In [215]:
gasolina_sp_2018.shape

(52, 23)

#### **Estatísticas Descritivas**

In [216]:
gasolina_sp_2018.describe()

,NÚMERO DE POSTOS PESQUISADOS,PREÇO MÉDIO REVENDA,DESVIO PADRÃO REVENDA,PREÇO MÍNIMO REVENDA,PREÇO MÁXIMO REVENDA,COEF DE VARIAÇÃO REVENDA,MÊS,ANO,coluna a partir de lista,PREÇO DE MEDIA (dolares)
count,52.000000,52.000000,52.000000,52.000000,52.000000,52.000000,52.000000,52.0,52.000000,52.000000
mean,1610.211538,4.202769,0.175500,3.671500,4.975750,0.041750,6.615385,2018.0,99196.096154,25.216615
std,213.627432,0.180672,0.017955,0.163808,0.211198,0.003965,3.470627,0.0,2279.595314,1.084032
min,102.000000,3.970000,0.150000,3.369000,4.499000,0.034000,1.000000,2018.0,95351.000000,23.820000
25%,1639.000000,4.008000,0.159750,3.499000,4.899000,0.039000,4.000000,2018.0,97281.250000,24.048000
50%,1642.000000,4.202000,0.173000,3.679000,4.999000,0.040000,7.000000,2018.0,99196.500000,25.212000
75%,1644.000000,4.336250,0.189000,3.784000,5.099000,0.044000,9.250000,2018.0,101116.250000,26.017500
max,1649.000000,4.512000,0.216000,3.969000,5.299000,0.051000,12.000000,2018.0,103025.000000,27.072000


In [218]:
gasolina_sp_2018['PREÇO MÉDIO REVENDA'].describe()

,PREÇO MÉDIO REVENDA
count,52.000000
mean,4.202769
std,0.180672
min,3.970000
25%,4.008000
50%,4.202000
75%,4.336250
max,4.512000


### 7.3 Como os preços da Gasolina Comum e do Etanol em São Paulo variaram em 2018?

In [224]:
 etanol_sp=default.query('(PRODUTO == "ETANOL HIDRATADO" or PRODUTO== "GASOLINA COMUM") and ESTADO == "SAO PAULO" and ANO== 2018' )

In [225]:
etanol_sp['PREÇO MÉDIO REVENDA'].describe().to_frame()

,PREÇO MÉDIO REVENDA
count,104.000000
mean,3.457337
std,0.766239
min,2.393000
25%,2.726000
50%,3.431000
75%,4.199000
max,4.512000


In [228]:
etanol_sp.groupby('PRODUTO')['PREÇO MÉDIO REVENDA'].describe()

,count,mean,std,min,25%,50%,75%,max
PRODUTO,,,,,,,,
ETANOL HIDRATADO,52.0,2.711904,0.141365,2.393,2.6315,2.720,2.83825,2.892
GASOLINA COMUM,52.0,4.202769,0.180672,3.970,4.0080,4.202,4.33625,4.512


<h2>8. Assuntos para continuar os estudos</h2>
<hr/>

- join
- concat
- plot
- data cleaning